### PR TITLE
Implement basic CLI client

### LINKED ISSUE #39: Implement vmd-client-cli
## Overview

We implement a program called `vmd-client-cli` which encapsulates client side access to vmd's REST API. This program will be used for Ghaf development, tests and CI as well as the `vmd-client` GUI-application. It acts as a layer that takes care of the mutual TLS connection between `vmd-server` and any client access such as a GUI-application or a shell script.

## Implementation proposal

- `vmd-client-cli` has a configuration file called `vmd-client.toml`. This file contains the configurations for client-cert, client-key, certificate-authority and base-URL. Config location can be set as environment variable `VMD_CLIENT_CONFIG_PATH=<path/>` or given as an argument using `--config <path/>`. All config options can be given as optional CLI parameters instead of a config file.
- `vmd-client-cli` takes a single argument as a path variable such as `vmd-client-cli vm/info/{ID}` and executes the request.
- Output formatting can have several options as flags such as `--json, --xml --deser`.
- By default outputs to stdout, but has optional `--output <path/>` argument.
- Timeout after 1 second if server hangs.

I propose that we implement the client with a simple CLI that takes in a path that corresponds to paths defined in the `vmd-api/openapi.yaml`. The REST API is well documented and building an additional API layer on top of it by enumerating all the API options in the CLI would be additional development overhead and possibly require significant refactoring when API is changed.

## Examples

```
> vmd-client-cli vm/list --json
{
    active-vms: [
        "1",
        "2",
    ]
}
```

### PR DESCRIPTION
**Pull request by juliuskoskela**
## Overview

Implement Nix flake build system and a CLI interface according to current openAPI.

```
Usage: vmd-client [OPTIONS] --hostname <HOSTNAME> --port <PORT> --cacert <CACERT> --key <KEY> --cert <CERT> --output <OUTPUT> <COMMAND>

Commands:
  list
  info
  action
  help    Print this message or the help of the given subcommand(s)

Options:
      --hostname <HOSTNAME>
  -p, --port <PORT>
      --cacert <CACERT>
      --key <KEY>
      --cert <CERT>
      --verbose
      --output <OUTPUT>      [possible values: json, yaml, text]
  -h, --help                 Print help
  -V, --version              Print version
  ```

## Requirements

- [x] Root flake.nix with separate outputs for client and server
- [x] Client CLI
- [x] JSON output
- [x] YAML output
- [x] TEXT output

## Testing (NixOS)

1. Open development shell

1. Create certificates

```bash
make -C test/auth
```

2. Start server

command:

```bash
nix run .#packages.x86_64-linux.vmd-server -- \
    --hostname localhost \
    --port 8080 \
    --cacert ./test/auth/certs/sample-ca-crt.pem \
    --cert ./test/auth/certs/sample-vmd-server-crt.pem \
    --key ./test/auth/certs/sample-vmd-server-key.pem
```

expected output:

```shell
⚡    VMD Server   ⚡

📌 api-version:           1.0.1
📡 address:               https://localhost:8080
🚩 base-path:             /api/v1
🔑 private-key:           ./test/auth/certs/sample-vmd-server-key.pem
🔐 certificate:           ./test/auth/certs/sample-vmd-server-crt.pem
🔐 certificate-authority: ./test/auth/certs/sample-ca-crt.pem
🔐 ocsp:                  🚫
```

3. Run client

command:

```bash
nix run .#packages.x86_64-linux.vmd-client -- \
    --hostname localhost \
    --port 8080 \
    --cacert test/auth/certs/sample-ca-crt.pem \
    --cert test/auth/certs/sample-vmd-client-chain.pem \
    --key test/auth/certs/sample-vmd-client-key.pem \
    --output yaml \
    list
```

expected output:

```shell
- 1
- 2
- 3
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -63,7 +63,7 @@ distclean: clean
 	make distclean -C test/auth
 	rm -rf $(RUST_SERVER_API)
 	rm -rf $(RUST_CLIENT_API)
-	rm openapitools.json
+	# rm openapitools.json
```
Comment (jkuro-tii) [inline Makefile]: I think it's better to delete the line. Unless commented out is still useful - a comment would be welcome.

**Comment by vilvo** [comment]
Tested and the `openapi-generator-cli generate` spits an error:
```
vmd (1b55bc6) [?] via 🦀 v1.69.0
❯ nix-shell
these 12 paths will be fetched (575.03 MiB download, 1939.71 MiB unpacked):
  /nix/store/22m1lj0k9n8lpxj449w2hvyi9qvg9ihr-ncurses-6.3-p20220507-man
  /nix/store/443r495zl5yra1svpm3ijgjkbpd2h2fy-rustc-1.62.1
  /nix/store/52gvz3frkihb39c0kxy7pg71xwnrgl4q-libpfm-4.11.0
  /nix/store/8ic6v37rpqql3w3mw9wj1mqviidmibi3-ncurses-6.3-p20220507-dev
  /nix/store/d5lxgpkm4cmv28jv1wnrlwxvmw6z14w6-llvm-14.0.6
  /nix/store/ga5d5hard379hvqi903wd0wqzbqwsxan-llvm-14.0.6-dev
  /nix/store/gp0p9lnpxsh6k7m2kqk1nx0s4mhqzfwh-openapi-generator-cli-6.0.1
  /nix/store/ivh5qhzhjv2vmcfq265gjr4fxcaw9hry-cargo-1.62.1
  /nix/store/jj87a94xsxdknzkgvh37193ldk2bzcmg-llvm-14.0.6-lib
  /nix/store/nv3709gm4bbfzv5bxk457s55l7rcd72k-openjdk-headless-17.0.3+7
  /nix/store/xmkvckbnqdz3ifwhmqgdwgyc6x1y06ab-hook
  /nix/store/z56cn68vv1wd8izxlg0v78xnxix45p6m-nss-cacert-3.80
copying path '/nix/store/z56cn68vv1wd8izxlg0v78xnxix45p6m-nss-cacert-3.80' from 'https://cache.nixos.org'...
copying path '/nix/store/xmkvckbnqdz3ifwhmqgdwgyc6x1y06ab-hook' from 'https://cache.nixos.org'...
copying path '/nix/store/jj87a94xsxdknzkgvh37193ldk2bzcmg-llvm-14.0.6-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/22m1lj0k9n8lpxj449w2hvyi9qvg9ihr-ncurses-6.3-p20220507-man' from 'https://cache.nixos.org'...
copying path '/nix/store/52gvz3frkihb39c0kxy7pg71xwnrgl4q-libpfm-4.11.0' from 'https://cache.nixos.org'...
copying path '/nix/store/nv3709gm4bbfzv5bxk457s55l7rcd72k-openjdk-headless-17.0.3+7' from 'https://cache.nixos.org'...
copying path '/nix/store/8ic6v37rpqql3w3mw9wj1mqviidmibi3-ncurses-6.3-p20220507-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/d5lxgpkm4cmv28jv1wnrlwxvmw6z14w6-llvm-14.0.6' from 'https://cache.nixos.org'...
copying path '/nix/store/ga5d5hard379hvqi903wd0wqzbqwsxan-llvm-14.0.6-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/443r495zl5yra1svpm3ijgjkbpd2h2fy-rustc-1.62.1' from 'https://cache.nixos.org'...
copying path '/nix/store/ivh5qhzhjv2vmcfq265gjr4fxcaw9hry-cargo-1.62.1' from 'https://cache.nixos.org'...
copying path '/nix/store/gp0p9lnpxsh6k7m2kqk1nx0s4mhqzfwh-openapi-generator-cli-6.0.1' from 'https://cache.nixos.org'...

vmd (1b55bc6) [?] via 🦀 v1.69.0 via ❄️  impure (nix-shell)
❯ make test-server-up
make -C test/auth
make[1]: Entering directory '/home/vilvo/vmd/test/auth'
mkdir -p certs
Generating CA key and certificate
Generating a RSA private key
..........................+++++
.........................+++++
writing new private key to 'certs/sample-ca-key.pem'
-----
echo "sh"
sh
Generating vmd-server key and certificate signing request
openssl genrsa -out certs/sample-vmd-server-key.pem 4096
Generating RSA private key, 4096 bit long modulus (2 primes)
................................................................................++++
.......................................++++
e is 65537 (0x010001)
openssl req -new \
	-key certs/sample-vmd-server-key.pem \
	-out certs/sample-vmd-server-csr.pem \
	-passin pass:1234 \
	-subj "/C=FI/ST=Uusimaa/L=Helsinki/O=VMD/OU=VMD/CN=localhost"
bash -c 'openssl x509 -req \
	-days 9999 \
	-in certs/sample-vmd-server-csr.pem \
	-CA certs/sample-ca-crt.pem \
	-CAkey certs/sample-ca-key.pem -CAcreateserial \
	-passin pass:1234 \
	-out certs/sample-vmd-server-crt.pem \
	-extfile <(printf "subjectAltName=DNS:localhost")'
Signature ok
subject=C = FI, ST = Uusimaa, L = Helsinki, O = VMD, OU = VMD, CN = localhost
Getting CA Private Key
openssl verify -CAfile certs/sample-ca-crt.pem certs/sample-vmd-server-crt.pem
certs/sample-vmd-server-crt.pem: OK
cat certs/sample-vmd-server-crt.pem certs/sample-vmd-server-key.pem > certs/sample-vmd-server-chain.pem
Generating client key and certificate signing request
openssl genrsa -out certs/sample-vmd-client-key.pem 4096
Generating RSA private key, 4096 bit long modulus (2 primes)
...................................................................................................................................................................................................................++++
.........++++
e is 65537 (0x010001)
openssl req -new \
	-key certs/sample-vmd-client-key.pem \
	-out certs/sample-vmd-client-csr.pem \
	-passin pass:1234 \
	-subj "/C=FI/ST=Uusimaa/L=Helsinki/O=VMD/OU=VMD/CN=vmd-client"
bash -c 'openssl x509 -req \
	-days 9999 \
	-in certs/sample-vmd-client-csr.pem \
	-CA certs/sample-ca-crt.pem \
	-CAkey certs/sample-ca-key.pem -CAcreateserial \
	-passin pass:1234 \
	-out certs/sample-vmd-client-crt.pem \
	-extfile <(printf "subjectAltName=DNS:localhost")'
Signature ok
subject=C = FI, ST = Uusimaa, L = Helsinki, O = VMD, OU = VMD, CN = vmd-client
Getting CA Private Key
openssl verify -CAfile certs/sample-ca-crt.pem certs/sample-vmd-client-crt.pem
certs/sample-vmd-client-crt.pem: OK
cat certs/sample-vmd-client-crt.pem certs/sample-vmd-client-key.pem > certs/sample-vmd-client-chain.pem
make[1]: Leaving directory '/home/vilvo/vmd/test/auth'
openapi-generator-cli generate \
	-g rust-server \
	-i vmd-api/openapi.yaml \
	-o vmd-api/rust-server \
	--additional-properties=packageName=vmd-rust-server-api
[main] ERROR i.s.parser.SwaggerCompatConverter - failed to read resource listing
com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'openapi': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (String)"openapi: 3.1.0
info:
  contact:
    email: nikita.bazulin@unikie.com
  description: OpenAPI specification for virtual machine management on Ghaf platform
  license:
    name: Apache License 2.0
    identifier: Apache-2.0
  title: Virtual Machine Management API of the Ghaf platform - OpenAPI
  version: 1.0.1
externalDocs:
  description: Find out more about the Ghaf platform
  url: https://github.com/tiiuae/ghaf
servers:
- description: The primary server
  url: https://localhost/api/v1
- descripti"[truncated 7610 chars]; line: 1, column: 8]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2391)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:745)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:2961)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:2002)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:802)
	at com.fasterxml.jackson.databind.ObjectMapper._readTreeAndClose(ObjectMapper.java:4703)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:3076)
	at io.swagger.parser.SwaggerCompatConverter.readResourceListing(SwaggerCompatConverter.java:209)
	at io.swagger.parser.SwaggerCompatConverter.read(SwaggerCompatConverter.java:122)
	at io.swagger.parser.SwaggerCompatConverter.readWithInfo(SwaggerCompatConverter.java:93)
	at io.swagger.parser.SwaggerParser.readWithInfo(SwaggerParser.java:44)
	at io.swagger.v3.parser.converter.SwaggerConverter.readLocation(SwaggerConverter.java:90)
	at io.swagger.parser.OpenAPIParser.readLocation(OpenAPIParser.java:16)
	at org.openapitools.codegen.config.CodegenConfigurator.toContext(CodegenConfigurator.java:573)
	at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput(CodegenConfigurator.java:631)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:457)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
[main] ERROR i.s.parser.SwaggerCompatConverter - failed to read resource listing
com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'openapi': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (String)"openapi: 3.1.0
info:
  contact:
    email: nikita.bazulin@unikie.com
  description: OpenAPI specification for virtual machine management on Ghaf platform
  license:
    name: Apache License 2.0
    identifier: Apache-2.0
  title: Virtual Machine Management API of the Ghaf platform - OpenAPI
  version: 1.0.1
externalDocs:
  description: Find out more about the Ghaf platform
  url: https://github.com/tiiuae/ghaf
servers:
- description: The primary server
  url: https://localhost/api/v1
- descripti"[truncated 7610 chars]; line: 1, column: 8]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2391)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:745)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:2961)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:2002)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:802)
	at com.fasterxml.jackson.databind.ObjectMapper._readTreeAndClose(ObjectMapper.java:4703)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:3076)
	at io.swagger.parser.SwaggerCompatConverter.readResourceListing(SwaggerCompatConverter.java:209)
	at io.swagger.parser.SwaggerCompatConverter.read(SwaggerCompatConverter.java:122)
	at io.swagger.parser.SwaggerCompatConverter.readWithInfo(SwaggerCompatConverter.java:93)
	at io.swagger.parser.SwaggerParser.readWithInfo(SwaggerParser.java:44)
	at io.swagger.v3.parser.converter.SwaggerConverter.readLocation(SwaggerConverter.java:90)
	at io.swagger.parser.OpenAPIParser.readLocation(OpenAPIParser.java:16)
	at org.openapitools.codegen.config.CodegenConfigurator.toContext(CodegenConfigurator.java:573)
	at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput(CodegenConfigurator.java:631)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:457)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
[main] INFO  o.o.codegen.DefaultGenerator - Generating with dryRun=false
[main] INFO  o.o.c.ignore.CodegenIgnoreProcessor - Output directory (/home/vilvo/vmd/vmd-api/rust-server) does not exist, or is inaccessible. No file (.openapi-generator-ignore) will be evaluated.
Exception in thread "main" java.lang.RuntimeException: Issues with the OpenAPI input. Possible causes: invalid/missing spec, malformed JSON/YAML files, etc.
	at org.openapitools.codegen.DefaultGenerator.generate(DefaultGenerator.java:868)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:465)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
make: *** [Makefile:29: vmd-api/rust-server] Error 1
```

**Comment by juliuskoskela** [comment]
Committed building with flake which should resolve reproducibility issues.

File: Makefile (hunk 1 above)
Comment (juliuskoskela) [reply Makefile]: Done

**Comment by vilvo** [comment]
Tested the flake build to work on `x86_64-linux` and `aarch64-linux`. I've got no `darwin` to test but let's have @mikatammi test on it when he is back as he wanted (or gave the flake example) with the `eachDefaultSystem` instead of ghaf-style where systems are limited to `x86_64-linux` and `aarch64-linux` - https://github.com/tiiuae/ghaf/blob/main/flake.nix#L45-L53 

If this is ready for review, please squash the commits. It keeps the version history in `main` more relevant to this implementation (not all nitty-gritty details on formatting and cleaning of your journey in this branch). If you squash (see `git rebase -i`) all the commits into one commit and sign the one commit, it will also fix the one commit missing the signing 27cadc0d381c5077a06129ca03d9dcc9250069cf  When done in your local branch, please force push over the PR branch and add reviewers.

**Review by vilvo — APPROVED** [review]
lgtm. Author wished to squash in github UI. Let's see how unsigned commit is managed in squashing.

**Pull request merged**